File: src/java/EntityFX/src/EntityFX/Core/Linpack/ParallelLinpackBenchmark.java
package EntityFX.Core.Linpack;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import EntityFX.Core.Writer;
import EntityFX.Core.Generic.BenchResult;
import EntityFX.Core.Generic.BenchmarkBase;

public class ParallelLinpackBenchmark extends BenchmarkBase<BenchResult[]> {

    public ParallelLinpackBenchmark(Writer writer, boolean printToConsole) throws FileNotFoundException {
        super(writer, printToConsole);
        Ratio = 5;
        isParallel = true;
    }

    @Override
    public BenchResult[] benchImplementation() throws InterruptedException, ExecutionException {
        return benchInParallel(() -> {
            try {
                return new Linpack(false);
            } catch (FileNotFoundException e1) {
                e1.printStackTrace();
            }
            return null;
        }, a -> {
            try {
                return a.run_benchmark(2000);
            } catch (IOException e) {
                e.printStackTrace();
            }
            return null;
        }, (a, r) -> {
            r.Points = a.MFLOPS * Ratio;
            r.Result = a.MFLOPS;
            r.Output = a.output;
        });
    }

    @Override
    public BenchResult populateResult(BenchResult benchResult, BenchResult[] linpackBenchResults) {
        BenchResult result = buildParallelResult(benchResult, linpackBenchResults);
        result.Result = Arrays.stream(linpackBenchResults).mapToDouble(r -> r.Result).sum();
        result.Units = "MFLOPS";
        result.Output = String.join("", Arrays.toString(Arrays.stream(linpackBenchResults).map(r -> r.Output).toArray()));
        return result;
    }

    @Override
    public void warmup(Double aspect) {
    }
}

File: src/python/entityfx/parallel_linpack_benchmark.py
from entityfx.linpack import Linpack
from entityfx.benchmark_base import BenchmarkBase
from entityfx.writer import Writer

class ParallelLinpackBenchmark(BenchmarkBase):

    def __init__(self, writer: Writer, print_to_console : bool=True, is_enabled : bool=True) -> None:
        super().__init__(writer, print_to_console, is_enabled)
        self.is_parallel = True
        self.ratio = 10
    
    def benchImplementation(self) -> list:
        return self.bench_in_parallel(lambda : Linpack(False), lambda a: a.bench(1000), lambda a, r : self.setBenchResultFunc(a, r))

    def warmup(self, aspect : float=.05) -> None:
        pass

    def populateResult(self, bench_result, results : list):
        result = self._buildParallelResult(bench_result, results) 
        result["Result"] = sum(map(lambda x : x["Result"], results))
        result["Units"] = "MFLOPS"
        result["Output"] = "".join(map(lambda x : x["Output"], results))
        return result

    def setBenchResultFunc(self, a, r):
        r["Points"] = a["MFLOPS"] * self.ratio
        r["Result"] = a["MFLOPS"]
        r["Output"] = a["Output"]